### PR TITLE
feat(HttpApp): Merge HttpEndPoint and HttpApp

### PR DESCRIPTION
**Pull request by amitksingh1490**

### CODE REVIEW

File: zio-http/src/test/scala/zhttp/experiment/HttpEndpointResponseSpec.scala [hunk 1]
```diff
@@ -49,6 +49,65 @@ object HttpEndpointResponseSpec extends DefaultRunnableSpec with HttpMessageAsse
             })
           }
         } +
+        testM("fromEffectFunction") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/experiment/HttpEndpointResponseSpec.scala]: I don't think we need these tests

**Review by tusharmath — CHANGES_REQUESTED** [review]

File: zio-http/src/main/scala/zhttp/experiment/HttpEndpoint.scala [hunk 2]
```diff
@@ -26,8 +26,8 @@ case class HttpEndpoint[-R, +E](http: Http[R, E, Request, Response[R, E]]) { sel
     evE: E <:< Throwable,
   ): ChannelHandler =
     new ChannelInboundHandlerAdapter { ad =>
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/experiment/HttpEndpoint.scala]: Can we merge `HttpEndpoint` with `HttpApp`

File: zio-http/src/main/scala/zhttp/http/package.scala [hunk 3]
```diff
@@ -7,9 +7,6 @@ import java.nio.charset.Charset
 
 package object http extends PathModule with RequestSyntax with HttpAppSyntax with RouteDecoderModule {
   type UHttp[-A, +B]      = Http[Any, Nothing, A, B]
-  type HttpApp[-R, +E]    = Http[R, E, Request, Response[R, E]]
-  type UHttpApp           = HttpApp[Any, Nothing]
-  type RHttpApp[-R]       = HttpApp[R, Throwable]
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/package.scala]: We should keep these aliases.

File: zio-http/src/test/scala/zhttp/experiment/internal/EndpointClient.scala [hunk 4]
```diff
@@ -4,7 +4,7 @@ import io.netty.buffer.Unpooled
 import io.netty.channel.embedded.EmbeddedChannel
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/experiment/internal/EndpointClient.scala:4]: can rename to `HttpAppClient`

File: zio-http/src/main/scala/zhttp/http/package.scala (hunk 3 above)
Comment (amitksingh1490) [reply zio-http/src/main/scala/zhttp/http/package.scala]: ```type HttpApp[-R, +E]    = Http[R, E, Request, Response[R, E]]```
was not needed others I have added

File: zio-http/src/test/scala/zhttp/experiment/internal/EndpointClient.scala (hunk 4 above)
Comment (amitksingh1490) [reply zio-http/src/test/scala/zhttp/experiment/internal/EndpointClient.scala:4]: Done

File: zio-http/src/main/scala/zhttp/experiment/HttpEndpoint.scala (hunk 2 above)
Comment (amitksingh1490) [reply zio-http/src/main/scala/zhttp/experiment/HttpEndpoint.scala]: Done

File: example/src/main/scala/Authentication.scala [hunk 5]
```diff
@@ -27,29 +27,28 @@ object Authentication extends App {
...
+  def authenticate[R, E](fail: HttpApp[R, E], success: JwtClaim => HttpApp[R, E]): HttpApp[R, E] =
     HttpApp.fromFunction {
       _.getHeader("X-ACCESS-TOKEN")
         .flatMap(header => jwtDecode(header.value.toString))
         .fold[HttpApp[R, E]](fail)(success)
     }
-  }
 
   // Http app that requires a JWT claim
-  def user(claim: JwtClaim): UHttpApp = Http.collect {
```
Comment (tusharmath) [inline example/src/main/scala/Authentication.scala:39]: can keep this type alias

File: example/src/main/scala/ConcreteEntity.scala [hunk 6]
```diff
@@ -17,9 +17,11 @@ object ConcreteEntity extends App {
...
     }
 
-  val app: Http[Any, Nothing, Request, Response[Any, Nothing]] = user
-    .contramap[Request](req => CreateUser(req.path.toString))   //Http[Any, Nothing, Request, UserCreated]
-    .map(userCreated => Response.text(userCreated.id.toString)) //Http[Any, Nothing, Request, Response]
+  val app: HttpApp[Any, Nothing] =
+    user
+      .contramap[Request](req => CreateUser(req.path.toString))   //Http[Any, Nothing, Request, UserCreated]
+      .map(userCreated => Response.text(userCreated.id.toString)) //Http[Any, Nothing, Request, Response]
+      .toHttpApp
```
Comment (tusharmath) [inline example/src/main/scala/ConcreteEntity.scala]:
```suggestion
      .toApp
```

File: example/src/main/scala/FileStreaming.scala [hunk 7]
```diff
@@ -19,5 +19,5 @@ object FileStreaming extends App {
 
   // Run it like any simple app
   override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
-    Server.start(8090, app.silent).exitCode
+    Server.start(8090, app).exitCode
```
Comment (tusharmath) [inline example/src/main/scala/FileStreaming.scala]: can we keep `silent` ?

File: zio-http-benchmarks/src/main/scala/zhttp.benchmarks/HttpRouteTextPerf.scala [hunk 8]
```diff
@@ -16,7 +16,7 @@ class HttpRouteTextPerf {
   private val res                        = Response.text("HELLO WORLD")
   private val app: HttpApp[Any, Nothing] = HttpApp.text("HELLO WORLD")
```
Comment (tusharmath) [inline zio-http-benchmarks/src/main/scala/zhttp.benchmarks/HttpRouteTextPerf.scala]: Write a benchmark for `Http` and not `HttpApp`

File: zio-http/src/main/scala/zhttp/http/CORS.scala [hunk 9]
```diff
@@ -80,11 +80,11 @@ object CORS {
               ),
             )
           case (_, Some(origin), _) if allowCORS(origin, req.method) =>
-            httpApp >>>
+            httpApp.http >>>
               Http.fromFunction[Response[R, E]](r => r.copy(headers = r.headers ++ corsHeaders(origin, req.method)))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/CORS.scala:84]:
```suggestion
              HttpApp.fromFunction[Response[R, E]](r => r.copy(headers = r.headers ++ corsHeaders(origin, req.method)))
```

File: zio-http/src/main/scala/zhttp/http/Http.scala [hunk 10]
```diff
@@ -256,6 +256,9 @@ sealed trait Http[-R, +E, -A, +B] extends (A => ZIO[R, Option[E], B]) { self =>
   final def provideSome[R1](r: R1 => R)(implicit ev: NeedsEnv[R]): Http[R1, E, A, B] =
     Http.fromPartialFunction[A](a => self(a).provideSome(r))
 
+  final def toHttpApp[R1 <: R, E1 >: E](implicit ev: Request <:< A, ev2: B <:< Response[R1, E1]): HttpApp[R1, E1] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Http.scala]:
```suggestion
  final def toApp[R1 <: R, E1 >: E](implicit evA: Request <:< A, evB: B <:< Response[R1, E1]): HttpApp[R1, E1] =
```

File: zio-http/src/main/scala/zhttp/http/Http.scala [hunk 11]
```diff
@@ -256,6 +256,9 @@ sealed trait Http[-R, +E, -A, +B] extends (A => ZIO[R, Option[E], B]) { self =>
...
+    HttpApp.mount(self.asInstanceOf[Http[R, E, Request, Response[R, E]]])
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Http.scala]:
```suggestion
    HttpApp(self.asInstanceOf[Http[R, E, Request, Response[R, E]]])
```

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Http.scala]: Remove `mount` from `HttpApp`

File: zio-http/src/main/scala/zhttp/http/HttpApp.scala [hunk 12]
```diff
@@ -1,72 +1,315 @@
...
+import io.netty.handler.codec.http._
+import zhttp.experiment.ContentDecoder
+import zhttp.http.HttpApp.InvalidMessage
+import zhttp.service.HttpRuntime
+import zio.stream.ZStream
+import zio.{Chunk, Promise, UIO, ZIO}
+
+import java.net.{InetAddress, InetSocketAddress}
+
+case class HttpApp[-R, +E](http: Http[R, E, Request, Response[R, E]]) { self =>
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpApp.scala]:
```suggestion
case class HttpApp[-R, +E](asHttp: Http[R, E, Request, Response[R, E]]) { self =>
```

File: zio-http/src/main/scala/zhttp/http/HttpApp.scala [hunk 13]
```diff
@@ -1,72 +1,315 @@
...
    */
-  def forbidden(msg: String): UHttpApp = Http.succeed(HttpError.Forbidden(msg).toResponse)
-
-  /**
-   * Creates a Http app from a function from Request to HttpApp
-   */
-  def fromFunction[R, E, B](f: Request => HttpApp[R, E]) = Http.fromFunction[Request](f)
-
+  def fromFunction[R, E, B](f: Request => HttpApp[R, E]): HttpApp[R, E] =
+    HttpApp(Http.fromFunction[Request](f(_).http).flatten)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpApp.scala]: Have consistent scala docs.

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 14]
```diff
@@ -40,7 +39,7 @@ object Server {
     maxRequestSize: Int = 4 * 1024, // 4 kilo bytes
     error: Option[Throwable => ZIO[R, Nothing, Unit]] = None,
     sslOption: ServerSSLOptions = null,
-    endpoint: HttpEndpoint[R, E] = HttpEndpoint.empty,
+    endpoint: HttpApp[R, E] = HttpApp.empty,
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
    app: HttpApp[R, E] = HttpApp.empty,
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 15]
```diff
@@ -50,10 +49,9 @@ object Server {
   private final case class Error[R](errorHandler: Throwable => ZIO[R, Nothing, Unit]) extends Server[R, Nothing]
   private final case class Ssl(sslOptions: ServerSSLOptions)                          extends UServer
   private final case class Address(address: InetSocketAddress)                        extends UServer
-  private final case class Endpoint[R, E](endpoint: HttpEndpoint[R, E])               extends Server[R, E]
+  private final case class Endpoint[R, E](endpoint: HttpApp[R, E])                    extends Server[R, E]
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  private final case class App[R, E](endpoint: HttpApp[R, E])                    extends Server[R, E]
```

File: zio-http/src/main/scala/zhttp/http/HttpApp.scala [hunk 16]
```diff
@@ -1,72 +1,315 @@
...
 
-  /**
-   * Creates an Http app from a function that returns a ZIO
-   */
-  def fromEffectFunction[R, E](f: Request => ZIO[R, E, Response[R, E]]): HttpApp[R, E] =
-    Http.fromEffectFunction(f)
+  final case class InvalidMessage(message: Any) extends IllegalArgumentException {
+    override def getMessage: String = s"Endpoint could not handle message: ${message.getClass.getName}"
+  }
+  def mount[R, E](http: Http[R, E, Request, Response[R, E]]): HttpApp[R, E] = HttpApp(http)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpApp.scala]:
```suggestion
  def fromHttp[R, E](http: Http[R, E, Request, Response[R, E]]): HttpApp[R, E] = HttpApp(http)
```

File: zio-http/src/main/scala/zhttp/http/HttpAppSyntax.scala [hunk 17]
```diff
@@ -8,11 +8,11 @@ private[zhttp] trait HttpAppSyntax {
...
   @deprecated("you no longer need to unwrap an HttpApp", "zio-http 1.0.0.0-RC18")
-  def asHttp: HttpApp[R, E] = self
+  def asHttp: Http[R, E, Request, Response[R, E]] = self.http
 
   /**
    * Converts a failing Http into a non-failing one by handling the failure and converting it to a result if possible.
    */
   def silent[R1 <: R, E1 >: E](implicit s: CanBeSilenced[E1, Response[R1, E1]]) =
-    self.catchAll(e => Http.succeed(s.silent(e)))
+    self.http.catchAll(e => Http.succeed(s.silent(e)))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpAppSyntax.scala]:
```suggestion
    self.catchAll(e => Http.succeed(s.silent(e)))
```

File: example/src/main/scala/HelloWorldWithCORS.scala [hunk 18]
```diff
@@ -14,5 +14,5 @@ object HelloWorldWithCORS extends App {
 
   // Run it like any simple app
   override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
-    Server.start(8090, app.silent).exitCode
+    Server.start(8090, app).exitCode
```
Comment (amitksingh1490) [inline example/src/main/scala/HelloWorldWithCORS.scala]: Keep Silent

File: example/src/main/scala/StickyThread.scala [hunk 19]
```diff
@@ -36,5 +36,5 @@ object StickyThread extends App {
 
   // Run it like any simple app
   override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
-    Server.start(8090, app.silent).exitCode
+    Server.start(8090, app).exitCode
```
Comment (amitksingh1490) [inline example/src/main/scala/StickyThread.scala]: keep silent

File: example/src/main/scala/StreamingResponse.scala [hunk 20]
```diff
@@ -28,6 +28,6 @@ object StreamingResponse extends App {
   override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = {
 
     // Starting the server (for more advanced startup configuration checkout `HelloWorldAdvanced`)
-    Server.start(8090, app.silent).exitCode
+    Server.start(8090, app).exitCode
```
Comment (amitksingh1490) [inline example/src/main/scala/StreamingResponse.scala]: keep silent

File: example/src/main/scala/ThreadSafety.scala [hunk 21]
```diff
@@ -19,5 +19,5 @@ object ThreadSafety extends App {
 
   // Run it like any simple app
   override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
-    Server.start(8090, app.silent).exitCode
+    Server.start(8090, app).exitCode
```
Comment (amitksingh1490) [inline example/src/main/scala/ThreadSafety.scala]: keep silent

File: example/src/main/scala/Authentication.scala (hunk 5 above)
Comment (amitksingh1490) [reply example/src/main/scala/Authentication.scala:39]: Done

File: zio-http/src/test/scala/zhttp/http/HttpSpec.scala [hunk 22]
```diff
@@ -37,6 +39,12 @@ object HttpSpec extends DefaultRunnableSpec with HExitAssertion {
         assert(actual)(isSuccess(equalTo("B")))
       },
     ),
+    suite("toHttpApp")(
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/HttpSpec.scala]:
```suggestion
    suite("toApp")(
```

**Pull request merged**